starboard/nplb/player_write_sample_test.cc

```diff
@@ -102,16 +103,176 @@ TEST_P(SbPlayerWriteSampleTest, WriteMultipleBatches) {
+    while (written_duration < buffer_duration) {
+      samples.AddAudioSamples(
+          i, 1, written_duration, written_duration,
+          std::max<int64_t>(
+              0, buffer_duration - written_duration - segment_duration));
+      written_duration += segment_duration;
+    }
+  }
+  samples.AddAudioEOS();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.Write(samples));
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerPresenting());
+
+  SbTime start_system_time = SbTimeGetMonotonicNow();
+  SbTime start_media_time = player_fixture.GetCurrentMediaTime();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerEndOfStream());
+
+  SbTime end_system_time = SbTimeGetMonotonicNow();
+  SbTime end_media_time = player_fixture.GetCurrentMediaTime();
+
+  const SbTime kDurationDifferenceAllowance = kSbTimeSecond;
+  ASSERT_NEAR(end_media_time, kDurationToPlay, kDurationDifferenceAllowance);
+  ASSERT_NEAR(end_system_time - start_system_time + start_media_time,
+              kDurationToPlay, kDurationDifferenceAllowance);
+}
+
+TEST_P(SbPlayerWriteSampleTest, PartialAudioDiscardAll) {
+  if (!IsPartialAudioSupported()) {
+    // TODO: Use GTEST_SKIP when we have a newer version of gtest.
+    SB_LOG(INFO)
+        << "The platform doesn't support partial audio. Skip the tests.";
+    return;
+  }
+
+  SbPlayerTestFixture player_fixture(GetParam());
+  if (HasFatalFailure()) {
+    return;
+  }
+  if (!player_fixture.HasAudio()) {
+    // TODO: Use GTEST_SKIP when we have a newer version of gtest.
+    SB_LOG(INFO) << "Skip PartialAudio test for audioless content.";
+    return;
+  }
+
+  const SbTime kDurationToPlay = kSbTimeSecond;
+  const SbTime kDurationPerWrite = 100 * kSbTimeMillisecond;
+  const SbTime kNumberOfBuffersToDiscard = 100;
+
+  GroupedSamples samples;
+  if (player_fixture.HasVideo()) {
+    samples.AddVideoSamples(
+        0, player_fixture.ConvertDurationToVideoBufferCount(kDurationToPlay));
+    samples.AddVideoEOS();
+  }
+
+  int written_buffer_index = 0;
+  SbTime written_duration = 0;
+  int num_of_buffers_per_write =
+      player_fixture.ConvertDurationToAudioBufferCount(kDurationPerWrite);
+  while (written_duration < kDurationToPlay) {
+    // Discard from front.
+    for (int i = 0; i < kNumberOfBuffersToDiscard; i++) {
+      samples.AddAudioSamples(written_buffer_index, 1, written_duration,
+                              kDurationPerWrite, 0);
+    }
+
+    samples.AddAudioSamples(written_buffer_index, num_of_buffers_per_write);
+    written_buffer_index += num_of_buffers_per_write;
+    written_duration += kDurationPerWrite;
+
+    // Discard from back.
+    for (int i = 0; i < kNumberOfBuffersToDiscard; i++) {
+      samples.AddAudioSamples(written_buffer_index, 1, written_duration, 0,
+                              kDurationPerWrite);
+    }
+  }
+  samples.AddAudioEOS();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.Write(samples));
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerPresenting());
```

Comment:
`player_fixture.Write()` uses a loop to write all samples.  Assume we are writing 100 samples, and after writing the first one, the `SbPlayer` sends `kSbPlayerStatePresenting`.  But the SbPlayer controls the pace of need data call, so the Write() call happens slowly while the SbPlayer is actually playing.
When the Write() finishes, there are about 0.5s of audio data inside the buffer, and we call `WaitForPlayerPresenting()` and start the timer.  Just want to double check if this behavior aligns to our expectation.